README.md
[![Build Status](https://travis-ci.org/Songmu/p5-Cache-Redis.png?branch=master)](https://travis-ci.org/Songmu/p5-Cache-Redis) [![Coverage Status](https://coveralls.io/repos/Songmu/p5-Cache-Redis/badge.png?branch=master)](https://coveralls.io/r/Songmu/p5-Cache-Redis?branch=master)
# NAME

Cache::Redis - Redis client specialized for cache

# SYNOPSIS

    use Cache::Redis;

    my $cache = Cache::Redis->new(
        server    => 'localhost:9999',
        namespace => 'cache:',
    );
    $cache->set('key', 'val');
    my $val = $cache->get('key');
    $cache->remove('key');



# DESCRIPTION

This module is for cache of Redis backend having [Cache::Cache](http://search.cpan.org/perldoc?Cache::Cache) like interface.

__THIS IS A DEVELOPMENT RELEASE. API MAY CHANGE WITHOUT NOTICE__.

# INTERFACE

## Methods

### `my $obj = Cache::Redis->new(%options)`

Create a new cache object. Various options may be set in `%options`, which affect
the behaviour of the cache (defaults in parentheses):

- `redis`

    Instance of Redis class are used as backend. If this is not passed, [Cache::Redis](http://search.cpan.org/perldoc?Cache::Redis) load from `redis_class` automatically.

- `redis_class ('Redis')`

    The class for backend.

- `default_expires_in (60*60*24 * 30)`

    The default expiration seconds for objects place in the cache.

- `namespace ('')`

    The namespace associated with this cache.

- `nowait (0)`

    If enabled, when you call a method that only returns its success status (like "set"), in a void context,
    it sends the request to the server and returns immediately, not waiting the reply. This avoids the
    round-trip latency at a cost of uncertain command outcome.

- `serializer ('Storable')`

    Serializer. 'MessagePack' and 'Storable' are usable. if \`serialize\_methods\` option
    is specified, this option is ignored.

- `serialize_methods (undef)`

    The value is a reference to an array holding two code references for serialization and
    de-serialization routines respectively.

- server (undef)

    Redis server information. You can use \`sock\` option instead of this and can specify
    all other [Redis](http://search.cpan.org/perldoc?Redis) constructor options to `Cache::Cache->new` method.

### `$obj->set($key, $value, $expire)`

Set a stuff to cache.

### `$obj->set_multi([$key, $value, $expire], [$key, $value])`

Set multiple stuffs to cache. stuffs is array reference.

### `my $stuff = $obj->get($key)`

Get a stuff from cache.

### `my $res = $obj->get_multi(@keys)`

Get multiple stuffs as hash reference from cache. `@keys` should be array.
A key is not stored on cache don't be contain `$res`.

### `$obj->remove($key)`

Remove stuff of key from cache.

### `$obj->get_or_set($key, $code, $expire)`

Get a cache value for _$key_ if it's already cached. If it's not cached then,
run _$code_ and cache _$expiration_ seconds and return the value.

### `$obj->nowait_push`

Wait all response from Redis. This is intended for `$obj->nowait`.

# DEPENDENCIES

Perl 5.8.1 or later.

# BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

# SEE ALSO

[perl](http://search.cpan.org/perldoc?perl)

# AUTHOR

Masayuki Matsuki <y.songmu@gmail.com>

# LICENSE AND COPYRIGHT

Copyright (c) 2013, Masayuki Matsuki. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension Cache::Redis

{{$NEXT}}

0.13 2015-07-13T10:08:42Z
    - fix argument check in `set()`

0.12 2015-07-07T12:54:14Z
    - update set to return redis response messages (dnmfarrell++)
    - `set` accepts callback (dnmfarrell++)

0.11 2015-04-15T03:47:42Z
    - update packaging
    - refactor tests

0.10 2014-03-28T08:05:19Z
    - Fix bug in get_multi

0.09 2014-03-28T06:39:09Z
    - Add redis option (soh335++)

0.08 2014-03-28T03:46:20Z
    - Add get_multi, set_multi method (soh335++)

0.07 2013-10-16T18:02:23Z
    - Add `redis_class` constructor option (EXPERIMENTAL)

0.06 2013-09-01T11:06:20Z
    - Add serializer `JSON`

0.05 2013-05-09T02:38:06Z
    - Do not run test case when redis-server command is not available

0.04 2013-05-07T10:31:33Z
    - fixup packaging and dependencies.

0.03 2013-05-07T06:44:39Z
    - Fix remove method bug. (didn't care namespace)
    - remove method returns removed value.
    - Switch to Minilla

0.02    Mon Apr  1 04:46:01 2013
    - default serializer turn to Storable **incompatible change**

0.01    Sun Mar 31 16:34:33 2013
    - original version


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Cache-Redis',
    module_name     => 'Cache::Redis',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


author/benchmark.pl
#!/usr/bin/env perl
use strict;
use warnings;
use utf8;

use Benchmark qw(:all);

use Cache::Memcached::Fast;
use Cache::Redis;
use File::Spec;
use File::Temp qw/tempdir/;
use Test::Memcached;
use Test::RedisServer;

my $tempdir = tempdir(CLEANUP => 1);
my $mem_sock = File::Spec->catfile($tempdir, 'mem.sock');
my $memd  = Test::Memcached->new(
    options => {
        unix_socket => $mem_sock,
    },
);
$memd->start;
my $memd_client = Cache::Memcached::Fast->new({
    servers => [$mem_sock],
});
my $redis = Test::RedisServer->new;
my $redis_client = Cache::Redis->new(
    $redis->connect_info,
    nowait => 1,
);
my $redis_client_fast = Cache::Redis->new(
    $redis->connect_info,
    redis_class => 'Redis::Fast',
    nowait => 1,
);
my $redis_client_hiredis = Cache::Redis->new(
    $redis->connect_info,
    redis_class => 'Redis::hiredis',
    nowait => 1,
);

my $results = timethese(0, {
    'memd'  => sub {
        $memd_client->set('hoge', 'fuga');
        $memd_client->get('hoge');
        $memd_client->remove('hoge');
    },
    'redis' => sub {
        $redis_client->set('hoge', 'fuga');
        $redis_client->get('hoge');
        $redis_client->remove('hoge');
    },
    'redis_fast' => sub {
        $redis_client_fast->set('hoge', 'fuga');
        $redis_client_fast->get('hoge');
        $redis_client_fast->remove('hoge');
    },
    'redis_hiredis' => sub {
        $redis_client_hiredis->set('hoge', 'fuga');
        $redis_client_hiredis->get('hoge');
        $redis_client_hiredis->remove('hoge');
    },
});
cmpthese( $results ) ;


lib/Cache/Redis.pm
package Cache::Redis;
use 5.008_001;
use strict;
use warnings;

our $VERSION = '0.12';

use Module::Load;

my $_mp;
sub _mp {
    $_mp ||= Data::MessagePack->new->utf8;
}
sub _mp_serialize {
    _mp->pack(@_);
}
sub _mp_deserialize {
    _mp->unpack(@_);
}

sub _mk_serialize {
    my $code = shift;

    return sub {
        my $data = shift;

        my $flags; # for future extention
        my $store_date = [$data, $flags];
        $code->($store_date);
    };
}

sub _mk_deserialize {
    my $code = shift;

    return sub {
        my $data = shift;

        my ($org, $flags) = @{$code->($data)};
        $org;
    };
}

sub new {
    my $class = shift;

    my $args = @_ == 1 ? $_[0] : {@_};
    my $default_expires_in = delete $args->{default_expires_in} || 60*60*24 * 30;
    my $namespace          = delete $args->{namespace}          || '';
    my $nowait             = delete $args->{nowait}             || 0;
    my $redis_class        = delete $args->{redis_class}        || 'Redis';

    my $redis             = delete $args->{redis};
    my $serializer        = delete $args->{serializer};
    my $serialize_methods = delete $args->{serialize_methods};
    die '`serializer` and `serialize_methods` is exclusive option' if $serializer && $serialize_methods;
    $serializer ||= 'Storable' unless $serialize_methods;

    my ($serialize, $deserialize);
    if ($serializer) {
        if ($serializer eq 'Storable') {
            require Storable;
            $serialize_methods = [\&Storable::nfreeze, \&Storable::thaw];
        }
        elsif ($serializer eq 'JSON') {
            require JSON::XS;
            $serialize_methods = [\&JSON::XS::encode_json, \&JSON::XS::decode_json];
        }
    }

    if ($serialize_methods) {
        $serialize   = _mk_serialize   $serialize_methods->[0];
        $deserialize = _mk_deserialize $serialize_methods->[1];
    }
    elsif ($serializer eq 'MessagePack') {
        require Data::MessagePack;
        $serialize   = \&_mp_serialize;
        $deserialize = \&_mp_deserialize;
    }
    die 'Serializer is not found' if !$serialize || !$deserialize;

    unless ( $redis ) {
        load $redis_class;
        $redis = $redis_class->new(
            encoding => undef,
            %$args
        );
    }

    bless {
        default_expires_in => $default_expires_in,
        serialize          => $serialize,
        deserialize        => $deserialize,
        redis              => $redis,
        namespace          => $namespace,
        nowait             => $nowait,
    }, $class;
}

sub get {
    my ($self, $key) = @_;
    $key = $self->{namespace} . $key;

    my $data = $self->{redis}->get($key);

    defined $data ? $self->{deserialize}->($data) : $data;
}

sub get_multi {
    my ($self, @keys) = @_;
    @keys = map { $self->{namespace} . $_ } @keys;

    my @data = $self->{redis}->mget(@keys);

    my $i = 0;
    my $ret = {};
    for my $key ( @keys ) {
        if ( defined $data[$i] ) {
            $ret->{$key} = $self->{deserialize}->($data[$i]);
        }
        $i++;
    }

    $ret;
}

sub set {
    my ($self, $key, $value, $expire, $callback) = @_;

    die 'set() requires key and value arguments' if scalar(@_) < 3;

    my $response = $self->_set($key, $value, $expire, $callback);

    # return now as the callabck will be called when it's done processing
    return if ($self->{nowait});
    $self->{redis}->wait_all_responses;
    return $response;
}

sub set_multi {
    my ($self, @items) = @_;

    for my $item ( @items ) {
        $self->_set(@$item);
    }
    $self->{redis}->wait_all_responses unless $self->{nowait};
}

sub _set {
    my ($self, $key, $value, $expire, $callback) = @_;
    $key = $self->{namespace} . $key;
    $expire ||= $self->{default_expires_in};

    if ($self->{nowait} && $callback) {
      $self->{redis}->setex($key, $expire, $self->{serialize}->($value), $callback);
    }
    else {
      $self->{redis}->setex($key, $expire, $self->{serialize}->($value));
    }
}

sub get_or_set {
    my ($self, $key, $code, $expire) = @_;

    my $data = $self->get($key);
    unless (defined $data) {
        $data = $code->();
        $self->set($key, $data, $expire);
    }
    $data;
}

sub remove {
    my ($self, $key) = @_;

    my $data = $self->get($key);
    $key = $self->{namespace} . $key;
    $self->{redis}->del($key);

    $data;
}

sub nowait_push {
    shift->{redis}->wait_all_responses;
}

1;
__END__

=head1 NAME

Cache::Redis - Redis client specialized for cache

=head1 SYNOPSIS

    use Cache::Redis;

    my $cache = Cache::Redis->new(
        server    => 'localhost:9999',
        namespace => 'cache:',
    );
    $cache->set('key', 'val');
    my $val = $cache->get('key');
    $cache->remove('key');


=head1 DESCRIPTION

This module is for cache of Redis backend having L<Cache::Cache> like interface.

B<THIS IS A DEVELOPMENT RELEASE. API MAY CHANGE WITHOUT NOTICE>.

=head1 INTERFACE

=head2 Methods

=head3 C<< my $obj = Cache::Redis->new(%options) >>

Create a new cache object. Various options may be set in C<%options>, which affect
the behaviour of the cache (defaults in parentheses):

=over

=item C<redis>

Instance of Redis class are used as backend. If this is not passed, L<Cache::Redis> load from C<redis_class> automatically.

=item C<redis_class ('Redis')>

The class for backend.

=item C<default_expires_in (60*60*24 * 30)>

The default expiration seconds for objects place in the cache.

=item C<namespace ('')>

The namespace associated with this cache.

=item C<nowait (0)>

If enabled, when you call a method that only returns its success status (like "set"), in a void context,
it sends the request to the server and returns immediately, not waiting the reply. This avoids the
round-trip latency at a cost of uncertain command outcome.

=item C<serializer ('Storable')>

Serializer. 'MessagePack' and 'Storable' are usable. if `serialize_methods` option
is specified, this option is ignored.

=item C<serialize_methods (undef)>

The value is a reference to an array holding two code references for serialization and
de-serialization routines respectively.

=item server (undef)

Redis server information. You can use `sock` option instead of this and can specify
all other L<Redis> constructor options to C<< Cache::Cache->new >> method.

=back

=head3 C<< $obj->set($key, $value, $expire) >>

Set a stuff to cache.

=head3 C<< $obj->set_multi([$key, $value, $expire], [$key, $value]) >>

Set multiple stuffs to cache. stuffs is array reference.

=head3 C<< my $stuff = $obj->get($key) >>

Get a stuff from cache.

=head3 C<< my $res = $obj->get_multi(@keys) >>

Get multiple stuffs as hash reference from cache. C<< @keys >> should be array.
A key is not stored on cache don't be contain C<< $res >>.

=head3 C<< $obj->remove($key) >>

Remove stuff of key from cache.

=head3 C<< $obj->get_or_set($key, $code, $expire) >>

Get a cache value for I<$key> if it's already cached. If it's not cached then,
run I<$code> and cache I<$expiration> seconds and return the value.

=head3 C<< $obj->nowait_push >>

Wait all response from Redis. This is intended for C<< $obj->nowait >>.

=head1 DEPENDENCIES

Perl 5.8.1 or later.

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=head1 SEE ALSO

L<perl>

=head1 AUTHOR

Masayuki Matsuki E<lt>y.songmu@gmail.comE<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2013, Masayuki Matsuki. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
